File: C C++/Sorting/MergeSort.c
#include<stdio.h>
#include<stdlib.h>

void merge(int arr1[],int arr2[],int arr[],int n,int m)
{
	int i,k,j;
	i=j=k=0;

	while((i<n)&&(j<m))
	{
		if(arr1[i]<arr2[j])
		{
			arr[k++]=arr1[i++];
		}
		else
		{
			arr[k++]=arr2[j++];
		}
	}

	while(i<n)
	{
		arr[k++]=arr1[i++];
	}

	while(j<m)
	{
		arr[k++]=arr2[j++];
	}
}

void mySort(int arr[],int size)
{
	if(size>1)
	{
		int n = size/2;
		int m = size-n;

		int *arr1 = (int*)malloc(n*sizeof(int));
		int *arr2 = (int*)malloc(m*sizeof(int));

		int k = 0;

		for(int i=0;i<n;i++)
		{
			arr1[i]=arr[k++];
		}

		for(int i=0;i<m;i++)
		{
			arr2[i]=arr[k++];
		}

		mySort(arr1,n);
		mySort(arr2,m);
		merge(arr1,arr2,arr,n,m);
	}
}

void main()
{
	int arr[]={123,2,2,1,1,-123,123,5111,-13123};

	printf("Original array is : ");

	for(int i=0;i<9;i++)
	{
		printf(" %d",arr[i]);
	}

	mySort(arr,9);

	printf("\nSorted array is : ");

	for(int i=0;i<9;i++)
	{
		printf(" %d",arr[i]);
	}
}

File: Java/IntelliJ/src/MergeSort.java
/**
 * Created by Balraj on 15-08-2016.
 */
public class MergeSort
{
    static void merge(int arr1[],int arr2[], int arr[])
    {
        int n = arr1.length;
        int m = arr2.length;

        int x,y,z;
        x = y = z = 0;

        while((x<n)&&(y<m))
        {
            if(arr1[x]<arr2[y])
            {
                arr[z++] = arr1[x++];
            }
            else
            {
                arr[z++] = arr2[y++];
            }
        }

        while(x<n)
        {
            arr[z++] = arr1[x++];
        }

        while(y<m)
        {
            arr[z++] = arr2[y++];
        }
    }


    static void msort(int arr[])
    {
        int size = arr.length;

        if(size>1)
        {
            int n = size/2;
            int m = size-n;

            int arr1[] = new int[n];
            int arr2[] = new int[m];

            int k = 0;

            for(int i=0;i<n;i++)
            {
                arr1[i] = arr[k++];
            }

            for(int i=0;i<m;i++)
            {
                arr2[i] = arr[k++];
            }

            msort(arr1);
            msort(arr2);
            merge(arr1,arr2,arr);
        }
    }

    public static void main(String[] args)
    {
        int arr[] = {123,345,123123,34,1,2,9};
        msort(arr);

        for(int i=0;i<arr.length;i++)
        {
            System.out.print(" "+arr[i]);
        }
    }
}
